armeria-backend/src/main/scala/sttp/client3/armeria/ArmeriaBackend.scala

```diff
@@ -0,0 +1,198 @@
+    body.map(Response(_, code, statusText, headers, Nil, request.onlyMetadata))
+  }
+
+  private def headersFromArmeria(headers: ResponseHeaders): Seq[Header] = {
+    @tailrec
+    def accumulate(i: util.Iterator[Entry[AsciiString, String]], acc: Seq[Header] = Nil): Seq[Header] =
+      if (i.hasNext) {
+        val header = i.next()
+        val accNext = if (isPseudoHeader(header)) acc else acc :+ Header(header.getKey.toString, header.getValue)
+        accumulate(i, accNext)
+      } else acc
+
+    accumulate(headers.iterator())
+  }
+
+  private def isPseudoHeader(header: Entry[AsciiString, String]) = header.getKey.startsWith(PseudoHeaderPrefix)
+
+  private def encode: (InputStream, String) => InputStream = {
+    case (is, "gzip")    => new GZIPInputStream(is)
+    case (is, "deflate") => new InflaterInputStream(is)
+    case (_, enc)        => throw new UnsupportedEncodingException(s"Unsupported encoding: $enc")
+  }
+
+  private lazy val bodyFromResponseAs = new BodyFromResponseAs[Future, InputStream, Nothing, Nothing] {
+    override protected def withReplayableBody(
+        response: InputStream,
+        replayableBody: Either[Array[Byte], SttpFile]
+    ): Future[InputStream] = Future(replayableBody match {
+      case Left(bytes) => new ByteArrayInputStream(bytes)
+      case Right(file) => new BufferedInputStream(new FileInputStream(file.toFile))
+    })
+
+    override protected def regularIgnore(response: InputStream): Future[Unit] = Future.unit
+
+    override protected def regularAsByteArray(response: InputStream): Future[Array[Byte]] = Future(
+      response.readAllBytes()
+    )
+
+    override protected def regularAsFile(response: InputStream, file: SttpFile): Future[SttpFile] =
+      Future(Try(FileHelpers.saveFile(file.toFile, response))).map(_ => file)
+
+    override protected def regularAsStream(response: InputStream): Future[(Nothing, () => Future[Unit])] =
+      Future.failed(StreamingNotSupportedException)
+
+    override protected def handleWS[T](
+        responseAs: WebSocketResponseAs[T, _],
+        meta: ResponseMetadata,
+        ws: Nothing
+    ): Future[T] = ws
+
+    override protected def cleanupWhenNotAWebSocket(response: InputStream, e: NotAWebSocketException): Future[Unit] =
+      Future(response.close())
+
+    override protected def cleanupWhenGotWebSocket(response: Nothing, e: GotAWebSocketException): Future[Unit] =
+      response
+  }
+
+  override def close(): Future[Unit] = Future.unit
+
+  override implicit def responseMonad: MonadError[Future] = new FutureMonad()(ec)
+}
+
+object ArmeriaBackend {
+  def apply()(implicit ec: ExecutionContext = ExecutionContext.global): SttpBackend[Future, Any] =
+    new FollowRedirectsBackend[Future, Any](new ArmeriaBackend()(ec))
+
+  private val PseudoHeaderPrefix = ":"
+  private val MultipartNotSupportedException = new IllegalArgumentException("Multipart body is not supported")
```

Comment:
I'd create instances of the exceptions at the use-site, then at least some part of the stack trace is useful

Reply:
ok